### ISSUE TITLE
ARM64 conditional compare (`CCMP`) instruction is not properly lifted

### ISSUE THREAD

**Issue by jonpalmisc** [Vector35/binaryninja-api#2674]
If I understand correctly, this code checks the value of `w1` against a handful of constants — effectively a big `if (w1 == X || w1 == Y || ...)` — but when lifted the actual constants are not visible, which seems like bad lifting / incorrect behavior.

See the following comparison between the disassembly and MLIL:

<img width="1120" alt="mlil" src="https://user-images.githubusercontent.com/41548458/135922907-5dd2693d-3fd3-447b-b5c3-f5712093118d.png">

[arm64_ccmp_lifting.bndb.zip](https://github.com/Vector35/binaryninja-api/files/7281259/arm64_ccmp_lifting.bndb.zip)

**Comment by yrp604** [comment]
tl;dr: the disassembler doesn't properly define `instr.setflags` for CCMP (and probably not CCMN?)

Neat, this is actually a disassembler bug. Dumping the steps I did to figure this out in case it's useful for other lifting bugs.

The reference manual for regular CMP defines it has an alias SUBS XZR, Xn, Xm -- basically subtract Xm from Xn, set the flags and discard the results. You can see this [here](https://developer.arm.com/documentation/dui0801/g/A64-General-Instructions/CMP--shifted-register-). The lift for this instruction is basically the same:

```
    case ARM64_CMP:
		il.AddInstruction(il.Sub(REGSZ_O(operand1),
					ILREG_O(operand1),
					ReadILOperand(il, operand2, REGSZ_O(operand1)), SETFLAGS));
		break;
```

We're pretty sure CMP gets lifted properly, so now let's look at CCMP:

This instructions reference is [here](https://developer.arm.com/documentation/dui0801/g/A64-General-Instructions/CCMP--register-). CCMP is very similar to CMP, it tests the 'z' and branches based on the result. If the 'z' flag is true, it performs a CMP. If it's false, it sets the four flag bits to values encoded in the instruction.

The code for this lift is [here](https://github.com/vector35/arch-arm64/blob/master/il.cpp#L1230-L1252):
```
	case ARM64_CCMP:
		{
			LowLevelILLabel trueCode, falseCode, done;

			il.AddInstruction(il.If(GetCondition(il, operand4.cond), trueCode, falseCode));

			il.MarkLabel(trueCode);
			il.AddInstruction(il.Sub(REGSZ_O(operand1),
						ILREG_O(operand1),
						ReadILOperand(il, operand2, REGSZ_O(operand1)), SETFLAGS));
			il.AddInstruction(il.Goto(done));

			il.MarkLabel(falseCode);
			il.AddInstruction(il.SetFlag(IL_FLAG_N, il.Const(0, (IMM_O(operand3) >> 3) & 1)));
			il.AddInstruction(il.SetFlag(IL_FLAG_Z, il.Const(0, (IMM_O(operand3) >> 2) & 1)));
			il.AddInstruction(il.SetFlag(IL_FLAG_C, il.Const(0, (IMM_O(operand3) >> 1) & 1)));
			il.AddInstruction(il.SetFlag(IL_FLAG_V, il.Const(0, (IMM_O(operand3) >> 0) & 1)));

			il.AddInstruction(il.Goto(done));

			il.MarkLabel(done);
		}
		break;
```

It's a little more complex, but it seems to match the instruction description from the manual. Looking at MLIL is a bit late to diagnose lifter bugs, so let's look at the lifted LLIL for the function:

![image](https://user-images.githubusercontent.com/8716677/135974090-3da52b48-3ae1-4bcc-b2b2-9dd864b48a81.png)

Something kinda weird stands out if you look at this long enough -- the sub from the CMP is different from the sub from the CCMP. Specifically, the sub from the CMP sets the flags (thus its displayed with a '{*}' next to it) while the sub from the CCMP is just a regular subtraction (e.g. `w0 - w1`). This is particularly strange because the lines for the subtraction and flag setting are identical between CMP and CCMP! We can further confirm this by looking at the LLIL SSA:

![image](https://user-images.githubusercontent.com/8716677/135974021-01acdc8c-3045-4408-be43-b39cf4512084.png)

Note here that flag:z can never get a value from the subtraction at 6.

Digging into the [definition](https://github.com/vector35/arch-arm64/blob/master/il_macros.h#L37) of `SETFLAGS` provides a hint for the the answer: `#define SETFLAGS (instr.setflags ? IL_FLAGWRITE_ALL : IL_FLAGWRITE_NONE)` -- what is `instr.setflags` for CCMP? Modifying the arch to dump this value show it as 0 -- thus the subtraction on the true path will not set the flags, and the resulting IL will be incorrect.

We can easily test this by setting `instr.setflags = 1` in the lifter. Doing so gives us far more sensible LLIL:
![image](https://user-images.githubusercontent.com/8716677/135973686-157b7521-d4d1-42ab-8102-e24522d86595.png)

Which turns into nicer HLIL :)
![image](https://user-images.githubusercontent.com/8716677/135973719-04daa62b-e77c-407b-b046-2c5a09e49e1e.png)

While this is a useful shortcut to test, this flag should be set in the disassembler itself.

**Comment by yrp604** [comment]
https://github.com/Vector35/arch-arm64/blob/staging/arm64test.py#L2157 :)
